### PR TITLE
Feature(BWS) - Enable Regtest for copay / bitpay wallets (BTC Only)

### PR DESCRIPTION
**Pull request by justinkook**
# Requirements 
- bitcore-node running on http://localhost:3000
- bws running locally on http://localhost:3232/bws/api
- mongod running
- copay running on port: 8100
- bitcoin-core running on regtest mode (blue icon logo)

> mongo topology crashes sometimes due to notifications being incompatible in a web browser

### bitcore.config.json
1.  Add testnet and regtest to bitcore.config.json. Testnet config must match regtest settings.
```
"testnet": {
          "chainSource": "p2p",
          "trustedPeers": [
            {
              "host": "127.0.0.1",
              "port": 20020
            }
          ],
          "rpc": {
            "host": "127.0.0.1",
            "port": 20021,
            "username": "bitpaytest",
            "password": "local321"
          }
        },
"regtest": {
          "chainSource": "p2p",
          "trustedPeers": [
            {
              "host": "127.0.0.1",
              "port": 20020
            }
          ],
          "rpc": {
            "host": "127.0.0.1",
            "port": 20021,
            "username": "bitpaytest",
            "password": "local321"
          }
        }
```
### bitcore-wallet-service/config.js
2. Point testnet to http://localhost:3000 in BWS/config.js and set regtestEnabled to true.
```
blockchainExplorerOpts: {
    btc: {
      livenet: {
        url: 'https://api.bitcore.io'
      },
      testnet: {
        // set url to http://localhost:3000 here
        url: 'http://localhost:3000',
        // set regtestEnabled to true here
        regtestEnabled: true
      }
    },
...
```
### packages/bitcore-wallet-service/lib/blockchainexplorers/v8.js
- Pull the enableRegtest flag from config
```
const config = require('../../config');
```

- Update v8network function to convert testnet to regtest in API request.

```
function v8network(bwsNetwork) {
  if (bwsNetwork == 'livenet') return 'mainnet';
  if (
    bwsNetwork == 'testnet' &&
    config.blockchainExplorerOpts.btc.testnet.regtestEnabled
  ) {
    return 'regtest';
  }
  return bwsNetwork;
}
```
### copay/app-template/index-template.html
1. Comment out content security meta tag in the ```<head>```
```
// <meta http-equiv="Content-Security-Policy" content="default-src 'self'  ... >
```

# Creating a wallet on regtest network

## Steps:

1. Set the wallet service URL to
```
http://localhost:3232/bws/api
```
2. Select Testnet by pressing the slider button.

screen shot 2019-03-06 at 10 50 29 am

# Testing on mobile
Requirements:
- Mobile phone and PC must be connected to the same internet
- PC desktop ip address for localhost

To find ip address for PC run:
```
// 127.0.0.1 is equal to localhost
ifconfig | grep "inet " | grep -v 127.0.0.1
```

1. Inside copay project root directory run:
```
npm run apply:copay
```

2. Enter PC ip address followed by port in the mobile phone browser:
```
10.10.11.73:8100
```

3. Set wallet service url to PC ip address /bws/api when creating a new wallet
```
http://10.10.11.73:3232/bws/api
```

### CODE REVIEW

File: packages/bitcore-wallet-service/config.js [hunk 1]
```diff
@@ -21,47 +21,47 @@ var config = {
...
   blockchainExplorerOpts: {
     btc: {
       livenet: {
-        url: 'https://api.bitcore.io',
+        url: 'https://api.bitcore.io'
       },
       testnet: {
-        url: 'https://api.bitcore.io',
-      },
+        // uncomment to point to local regtest
```
Comment (nitsujlangston) [inline packages/bitcore-wallet-service/config.js]: remove comment

File: packages/bitcore-wallet-service/lib/blockchainexplorers/v8.js [hunk 2]
```diff
@@ -20,187 +20,187 @@ var Constants = Common.Constants,
   Defaults = Common.Defaults,
   Utils = Common.Utils;
 
-
 function v8network(bwsNetwork) {
   if (bwsNetwork == 'livenet') return 'mainnet';
+  // Uncomment to turn on regtest
+  // if (bwsNetwork == 'testnet') return 'regtest';
```
Comment (nitsujlangston) [inline packages/bitcore-wallet-service/lib/blockchainexplorers/v8.js]: make this pull from config `regtestEnabled: true` where would be appropriate in config

File: packages/bitcore-wallet-service/lib/blockchainexplorers/v8.js [hunk 3]
```diff
@@ -15,192 +15,196 @@ var Bitcore_ = {
...
   Defaults = Common.Defaults,
   Utils = Common.Utils;
 
-
 function v8network(bwsNetwork) {
   if (bwsNetwork == 'livenet') return 'mainnet';
+  if (
+    bwsNetwork == 'testnet' &&
+    config.blockchainExplorerOpts.btc.testnet.regtestEnabled
+  )
```
Comment (justinkook) [inline packages/bitcore-wallet-service/lib/blockchainexplorers/v8.js]: Conditional if network is testnet and regtestEnabled is true, change testnet to regtest

File: packages/bitcore-wallet-service/config.js [hunk 4]
```diff
@@ -21,47 +21,46 @@ var config = {
...
   blockchainExplorerOpts: {
     btc: {
       livenet: {
-        url: 'https://api.bitcore.io',
+        url: 'https://api.bitcore.io'
       },
       testnet: {
         url: 'https://api.bitcore.io',
-      },
+        regtestEnabled: false
```
Comment (justinkook) [inline packages/bitcore-wallet-service/config.js]: For internal testing, change url to localhost:3000 and set regtestEnabled to true

File: packages/bitcore-wallet-service/README.md [hunk 5]
```diff
@@ -70,211 +70,329 @@ BWS can be used with PM2 with the provided `app.js` script:
 ```
 
 @dabura667 made a report about how to use letsencrypt with BWS: https://github.com/bitpay/bitcore-wallet-service/issues/423
-  
+
+# Enabling Regtest Mode
```
Comment (micahriggan) [inline packages/bitcore-wallet-service/README.md]: Can you reset --hard from master and only pull in this change? There are quite a few files changed in this MR, and I feel like this README change is really the only thing needed

**Pull request closed without merging**